eventlog/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,293 @@
+// initialization once, subsequently returning the result of that attempt.
+func (c *CloudWatchEventer) Init(ctx context.Context) error {
+	c.initOnce.Do(func() {
+		defer func() {
+			if c.initErr != nil {
+				log.Error.Printf("CloudWatchEventer: failed to initialize event log: %v", c.initErr)
+			}
+		}()
+		var err error
+		_, err = c.client.CreateLogGroupWithContext(ctx, &cloudwatchlogs.CreateLogGroupInput{
+			LogGroupName: aws.String(c.group),
+		})
+		if err != nil {
+			aerr, ok := err.(awserr.Error)
+			if !ok || aerr.Code() != cloudwatchlogs.ErrCodeResourceAlreadyExistsException {
+				c.initErr = errors.E(fmt.Sprintf("could not create CloudWatch log group %s", c.group), err)
+				return
+			}
+		}
+		_, err = c.client.CreateLogStreamWithContext(ctx, &cloudwatchlogs.CreateLogStreamInput{
+			LogGroupName:  aws.String(c.group),
+			LogStreamName: aws.String(c.stream),
+		})
+		if err != nil {
+			aerr, ok := err.(awserr.Error)
+			if ok && aerr.Code() != cloudwatchlogs.ErrCodeResourceAlreadyExistsException {
+				c.initErr = errors.E(fmt.Sprintf("could not create CloudWatch log stream %s", c.stream), err)
+				return
+			}
+		}
+	})
+	return c.initErr
+}
+
+// sync syncs all buffered events to CloudWatch. This is mostly useful for
+// testing.
+func (c *CloudWatchEventer) sync() {
+	c.syncc <- struct{}{}
+	<-c.syncDonec
+}
+
+func (c *CloudWatchEventer) Close() error {
+	c.cancel()
+	<-c.donec
+	return nil
+}
+
+func (c *CloudWatchEventer) loop(ctx context.Context) {
+	var (
+		syncTimer      = time.NewTimer(syncInterval)
+		inputLogEvents []*cloudwatchlogs.InputLogEvent
+		batchSize      int
+	)
+	sync := func(drain bool) {
+		defer func() {
+			inputLogEvents = nil
+			batchSize = 0
+			if !syncTimer.Stop() && drain {
+				<-syncTimer.C
+			}
+			syncTimer.Reset(syncInterval)
+		}()
+		if len(inputLogEvents) == 0 {
+			return
+		}
+		if err := c.Init(ctx); err != nil {
+			return
+		}
+		response, err := c.client.PutLogEventsWithContext(ctx, &cloudwatchlogs.PutLogEventsInput{
+			LogEvents:     inputLogEvents,
+			LogGroupName:  aws.String(c.group),
+			LogStreamName: aws.String(c.stream),
+			SequenceToken: c.sequenceToken,
+		})
+		if err != nil {
+			log.Error.Printf("CloudWatchEventer: PutLogEvents error: %v", err)
+			if aerr, ok := err.(cloudwatchlogs.InvalidSequenceTokenException); ok {
+				c.sequenceToken = aerr.ExpectedSequenceToken
+			}
+			return
+		}
+		c.sequenceToken = response.NextSequenceToken
+	}
+	process := func(e event) {
+		s, err := marshal.Marshal(e.typ, e.fieldPairs)
+		if err != nil {
+			log.Error.Printf("CloudWatchEventer: dropping log event: %v", err)
+			return
+		}
+		if len(s) > maxSingleMessageSize {
+			log.Error.Printf("CloudWatchEventer: dropping log event: message too large")
+			return
+		}
+		newBatchSize := batchSize + len(s) + 26
+		if newBatchSize > maxBatchSize {
+			sync(true)
+		}
+		inputLogEvents = append(inputLogEvents, &cloudwatchlogs.InputLogEvent{
```

Comment:
It looks like this is all one thread and doesn't require any locking, right?

Reply:
That's right; there is only one goroutine accessing `inputLogEvents`, the one running `loop`.